src/host/selection.cpp

```diff
@@ -43,11 +43,28 @@ void Selection::_RegenerateSelectionSpans() const
     endSelectionAnchor.x = (_d->coordSelectionAnchor.x == _d->srSelectionRect.left) ? _d->srSelectionRect.right : _d->srSelectionRect.left;
     endSelectionAnchor.y = (_d->coordSelectionAnchor.y == _d->srSelectionRect.top) ? _d->srSelectionRect.bottom : _d->srSelectionRect.top;
 
+    // GH #18106: Conhost uses an inclusive range for selection,
+    //    but GetTextSpans() needs an exclusive range.
     const auto blockSelection = !IsLineSelection();
-    _lastSelectionSpans = screenInfo.GetTextBuffer().GetTextSpans(_d->coordSelectionAnchor,
-                                                                  endSelectionAnchor,
-                                                                  blockSelection,
-                                                                  false);
+    const auto& buffer = screenInfo.GetTextBuffer();
+    auto startSelectionAnchor = _d->coordSelectionAnchor;
+    if (blockSelection)
+    {
+        auto& targetEndpoint = startSelectionAnchor.x <= endSelectionAnchor.x ? endSelectionAnchor : startSelectionAnchor;

```

Comment:
Ah! I think this is similar to the thread on the Pivot test below.

Reverting the changes to this file causes the following behavior:
- drag right: exclusive end!
- drag left (aka pivot): inclusive end!
This is the case regardless of block selection.

Now, let's walk through the change (it's past 5 PM on a Friday and I'm thinking out loud at this point haha):
- to fix the issue above, let's do L61:
```
buffer.GetSize().IncrementInExclusiveBounds(startSelectionAnchor <= endSelectionAnchor ? endSelectionAnchor : startSelectionAnchor);
```
That fixes line selection 🎉. But now block selection is wonky as such:
- ✅ selecting towards the right & down works as expected
- ✅ selecting towards the left & up works as expected
- ❌ selecting towards the left & down --> "end" (left) is unnecessarily incremented & "start" (right) is unnecessarily decremented
- ❌ selecting towards the right & up --> "end" (right) is unnecessarily decremented & "start" (left) is unnecessarily incremented
Not exactly sure where the problematic behavior is coming from, but I can look into that later, if necessary.

With the current fix, that doesn't happen. We're making adjustments to specifically fix the problematic behavior.

Actually, I'm going to look into this further on Monday and add a doc comment. Leaving thread unresolved.

Reply:
Simplified in latest commit. Added a comment.

```cpp
    // GH #18106: Conhost and Terminal share most of the selection code.
    //    Both now store the selection data as a half-open range [start, end),
    //     where "end" is the bottom-right-most point.
    //    Note that Conhost defines start/end as "start was set in time before end",
    //     whereas above (and in Terminal) we're treating start/end as "start is physically before end".
    //    We want Conhost to still operate as an inclusive range.
    //    To make it "feel" inclusive, we need to adjust the "end" endpoint
    //     by incrementing it by one, so that the "end" endpoint is rendered
    //     and handled as selected.
    const auto blockSelection = !IsLineSelection();
    const auto& buffer = screenInfo.GetTextBuffer();
    auto startSelectionAnchor = _d->coordSelectionAnchor;
    buffer.GetSize().IncrementInExclusiveBounds(startSelectionAnchor.x <= endSelectionAnchor.x ? endSelectionAnchor : startSelectionAnchor);
    _lastSelectionSpans = buffer.GetTextSpans(startSelectionAnchor,
                                              endSelectionAnchor,
                                              blockSelection,
                                              false);
    _lastSelectionGeneration = _d.generation();
```